app/src/main/java/org/asteroidos/sync/ble/MediaService.java

```diff
@@ -96,6 +105,42 @@ public void unsync() {
         }
     }
 
+    private ContentObserver mVolumeChangeObserver = new ContentObserver(new Handler()) {
+        // The last value of volume send to the watch.
+        private int reportedVolume;
+        private Handler handler = new Handler();
+
+        private void updateVolume() {
+            if (mMediaController != null && mMediaController.getPlaybackInfo() != null) {
+                int vol = (100 * mMediaController.getPlaybackInfo().getCurrentVolume()) / mMediaController.getPlaybackInfo().getMaxVolume();
+
+                if (reportedVolume != vol) {
+                    reportedVolume = vol;
+                    // Set real volume.
+                    mVolume = reportedVolume;
+
+                    byte[] data = new byte[1];
+                    data[0] = (byte) mVolume;
+                    mDevice.write(mediaVolumeCharac, data, MediaService.this);
+                }
+            }
+        }
+        @Override
+        public void onChange(boolean selfChange) {
+            super.onChange(selfChange);
+            // Volume changes are reported to the watch. Sending updates to the watch while the user
+            // is changing the volume results in weird behaviour(lowering volume while user is incrementing volume)
+            // Delayed sending volume updates to the watch solves this issue. We do need to keep track of the (theoretical) volume manually.
+            handler.removeCallbacksAndMessages(null);
+            handler.postDelayed(new Runnable() {
+                @Override
+                public void run() {
+                    updateVolume();
+                }
+            }, 500);
```

Comment:
Doesn't this feel laggy when using the watch?
I would solve this differently by always having the watch show the real phone volume (for example, if you press "volume up", the asteroid-music logic would take a bit of time to update.

Reply:
Yes, it does feel slow to some degree. There are multiple points where this comes in.
- The code uses a `ContentObserver`. Which also seems slow to respond to volume changes.
- Then there is this added code to workaround the issue described in the second example of the pr.

I agree that we want to update the volume on the watch as quick as possible, to avoid the feeling of it lagging behind. But this add the issues described in the second example. There are other workarounds that may be possible that fix this:
- also send the old volume from asteroid-btsyncd
- only send increment/decrement from asteroid-btsyncd (this makes the implementation on AsteroidSync also much easier). But this makes the value set in Asteroid-music irrelevant (another nasty hack)

If you have another method to workaround this issues then it will be more then welcome 🙂. All methods I came up with seem non-optimal...

Reply:
I would indeed solve this with increment/decrement. asteroid-music would probably need to communicate over dbus directly with asteroid-btsyncd as I don't think MPRIS has an API for that, this promises to be a bit hacky but that's ok.
And then AsteroidOSSync would react to the volume up/down request, wait until its contentobserver updates, send the new volume back to asteroid-btsyncd which would either expose it over MPRIS or the custom DBus service we create... I guess MPRIS works well here.

The value of this is that we always show the real volume so we know that things are always in sync. Even though it takes a bit of time to update, that's fine, this is similar to pressing the next button where things also takes time. Is there an alternative to using ContentObserver that could potentially use less time?

What do you think?

Reply:
I have solved this specific issue by letting the delay be handled in the `asteroid-music` app. It now no longer feels slow as the volume bar is updating as the volume changes, even when the user is pressing the buttons rapidly.

The only issue I can think of is that the complexity of the code is increased in the `asteroid-music` app, as the volume value is buffered/delayed on that end to keep track of the 'theoretical' volume.